src/Utilities/functions.jl

```diff
@@ -1732,8 +1732,24 @@ function operate(
     return operate!(op, T, copy(f), g)
 end
 
+_eltype(args::Tuple) = _eltype(first(args), Base.tail(args))
+_eltype(::Tuple{}) = nothing
+_eltype(::MOI.Utilities.TypedScalarLike{T}, tail) where {T} = T
+_eltype(::MOI.VariableIndex, tail) = _eltype(tail)
+_eltype(::MOI.Utilities.TypedVectorLike{T}, tail) where {T} = T
+_eltype(::MOI.VectorOfVariables, tail) = _eltype(tail)
+
 # To avoid type piracy, we add at least one `ScalarLike` outside of the `...`.
-function Base.:+(arg::ScalarLike{T}, args::ScalarLike{T}...) where {T}
+function Base.:+(arg::ScalarLike, args::ScalarLike...)
```

Comment:
Okay. The problem here (and others below) is that `ScalarLike{T} = Union{MOI.VariableIndex,MOI.ScalarAffineFunction{T}`. 

The problem is that if you pass in one or more `MOI.VariableIndex`, then `T` is not defined. We previously worked around this by defining another fallback, but things get a bit confusing.

I'd say that writing methods like this works, but is probably bad practice. Unfortunately, I don't think there's a way to encode `T` using dispatch, so we have to use a look-up function `_eltype` internally. 

`_eltype` uses the same lispy recursion to extra type parameters.

The downside to this approach is that
```Julia
julia> x = MOI.VariableIndex(1)
MathOptInterface.VariableIndex(1)

julia> y = 1 * x
MathOptInterface.ScalarAffineFunction{Int64}(MathOptInterface.ScalarAffineTerm{Int64}[MathOptInterface.ScalarAffineTerm{Int64}(1, MathOptInterface.VariableIndex(1))], 0)

julia> z = 1.0 * x
MathOptInterface.ScalarAffineFunction{Float64}(MathOptInterface.ScalarAffineTerm{Float64}[MathOptInterface.ScalarAffineTerm{Float64}(1.0, MathOptInterface.VariableIndex(1))], 0.0)

julia> x + y + z
ERROR: MethodError: no method matching operate!(::typeof(+), ::Type{Int64}, ::MathOptInterface.ScalarAffineFunction{Int64}, ::MathOptInterface.ScalarAffineFunction{Float64})
Closest candidates are:
  operate!(::typeof(+), ::Type{T}, ::Any, ::Any, ::Any, ::Any...) where T at /Users/oscar/.julia/dev/MathOptInterface/src/Utilities/functions.jl:1558
  operate!(::Function, ::Type{T}, ::Union{AbstractVector{T}, T}...) where T at /Users/oscar/.julia/dev/MathOptInterface/src/Utilities/functions.jl:1186
  operate!(::Union{typeof(+), typeof(-)}, ::Type{T}, ::MathOptInterface.VariableIndex, ::Union{MathOptInterface.VariableIndex, MathOptInterface.ScalarAffineFunction{T}, MathOptInterface.ScalarQuadraticFunction{T}, T} where T) where T at /Users/oscar/.julia/dev/MathOptInterface/src/Utilities/functions.jl:1581
  ...
Stacktrace:
 [1] operate(::typeof(+), ::Type{Int64}, ::MathOptInterface.VariableIndex, ::MathOptInterface.ScalarAffineFunction{Int64}, ::MathOptInterface.ScalarAffineFunction{Float64})
   @ MathOptInterface.Utilities ~/.julia/dev/MathOptInterface/src/Utilities/functions.jl:1618
 [2] +(::MathOptInterface.VariableIndex, ::MathOptInterface.ScalarAffineFunction{Int64}, ::MathOptInterface.ScalarAffineFunction{Float64})
   @ MathOptInterface.Utilities ~/.julia/dev/MathOptInterface/src/Utilities/functions.jl:1753
 [3] top-level scope
   @ REPL[7]:1
```

But we already had a similar method error:
```Julia
julia> x = MOI.VariableIndex(1)
MathOptInterface.VariableIndex(1)

julia> y = 1 * x
MathOptInterface.ScalarAffineFunction{Int64}(MathOptInterface.ScalarAffineTerm{Int64}[MathOptInterface.ScalarAffineTerm{Int64}(1, MathOptInterface.VariableIndex(1))], 0)

julia> z = 1.0 * x
MathOptInterface.ScalarAffineFunction{Float64}(MathOptInterface.ScalarAffineTerm{Float64}[MathOptInterface.ScalarAffineTerm{Float64}(1.0, MathOptInterface.VariableIndex(1))], 0.0)

julia> x + y
MathOptInterface.ScalarAffineFunction{Int64}(MathOptInterface.ScalarAffineTerm{Int64}[MathOptInterface.ScalarAffineTerm{Int64}(1, MathOptInterface.VariableIndex(1)), MathOptInterface.ScalarAffineTerm{Int64}(1, MathOptInterface.VariableIndex(1))], 0)

julia> x + y + z
ERROR: MethodError: no method matching +(::MathOptInterface.ScalarAffineFunction{Int64}, ::MathOptInterface.ScalarAffineFunction{Float64})
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...) at operators.jl:560
  +(::Union{MathOptInterface.ScalarAffineFunction{T}, MathOptInterface.ScalarQuadraticFunction{T}}, ::T) where T at /Users/oscar/.julia/packages/MathOptInterface/lRbIA/src/Utilities/functions.jl:1752
  +(::MutableArithmetics.Zero, ::Any) at /Users/oscar/.julia/packages/MutableArithmetics/UtY4H/src/rewrite.jl:54
  ...
Stacktrace:
 [1] +(::MathOptInterface.VariableIndex, ::MathOptInterface.ScalarAffineFunction{Int64}, ::MathOptInterface.ScalarAffineFunction{Float64})
   @ Base ./operators.jl:560
 [2] top-level scope
   @ REPL[19]:1
```

One thing I could do is check in `_eltype` that all types are equivalent instead of returning the first one.

Reply:
Yes, it does not follow the methoderror principle but it should be ok